compiler/src/dotty/tools/dotc/transform/Splicing.scala

```diff
@@ -0,0 +1,409 @@
+      }
+    }
+
+    /** Transforms a splice at level 1 into a hole
+     *
+     *  Finds all terms and types that are defined in the current quote and used within this splice.
+     *  The resulting hole will contain all of these terms and types as arguments.
+     *  Note that these captured variables are stage correct.
+     *
+     *  For a `x` of type `T1` and a type `X` defined in the current quote
+     *  ```scala
+     *  ${  (using Quotes) ?=> {... x ... X ...} }: T2
+     *  ```
+     *  is transformed into
+     * ```scala
+     *  {{{ <holeIdx++> | T2 | x, X | (x$1: Expr[T1], X$1: Type[X]) => (using Quotes) ?=> {... ${x$1} ...  X$1.Underlying ...}  }}}
+     *  ```
+     */
+    private class SpliceTransformer(spliceOwner: Symbol) extends Transformer:
+      private var refBindingMap = mutable.Map.empty[Symbol, (Tree, Symbol)]
+      /** Reference to the `Quotes` instance of the current level 1 splice */
+      private var quotes: Tree = null
+      private var healedTypes: PCPCheckAndHeal.QuoteTypeTags = null
+
+      def transformSplice(tree: tpd.Tree, tpe: Type)(using Context): tpd.Tree =
+        assert(level == 0)
+        val newTree = transform(tree)
+        val (refs, bindings) = refBindingMap.values.toList.unzip
+        val bindingsTypes = bindings.map(_.termRef.widenTermRefExpr)
+        val methType = MethodType(bindingsTypes, newTree.tpe)
+        val meth = newSymbol(spliceOwner, nme.ANON_FUN, Synthetic | Method, methType)
+        val ddef = DefDef(meth, List(bindings), newTree.tpe, newTree.changeOwner(ctx.owner, meth))
+        val fnType = defn.FunctionType(bindings.size, isContextual = false).appliedTo(bindingsTypes :+ newTree.tpe)
+        val closure = Block(ddef :: Nil, Closure(Nil, ref(meth), TypeTree(fnType)))
+        holeIdx += 1
+        tpd.Hole(true, holeIdx, refs, closure, TypeTree(tpe))
+
+      override def transform(tree: tpd.Tree)(using Context): tpd.Tree =
+        tree match
+          case tree: RefTree =>
+            if tree.isTerm then
+              if quotedDefs.contains(tree.symbol) then
+                val tpe = tree.tpe.widenTermRefExpr match {
+                  case tpw: MethodicType => tpw.toFunctionType(isJava = false)
+                  case tpw => tpw
+                }
+                spliced(tpe)(capturedTerm(tree))
+              else super.transform(tree)
+            else // tree.isType then
+              if containsCapturedType(tree.tpe) then
+                val capturedTypeSym = capturedType(tree)
+                if level >= 1 then
+                  TypeTree(healedTypes.getTagRef(capturedTypeSym.termRef))
+                else
+                  ref(capturedTypeSym).select(defn.QuotedType_splice)
+              else super.transform(tree)
+          case tree: TypeTree =>
+            if containsCapturedType(tree.tpe) then
+              // TODO factor out the common code with the above case
+              val capturedTypeSym = capturedType(tree)
```

Comment:
Yes. For example in the following quote 

```scala
'{ type X; type Y; ${ ... f[Map[X, Y]] ...  } }
```
we would end up with 

```scala
'{ type X; type Y; {{{ 0 | T | Map[X, Y] | (t: Type[Map[X, Y]]) => ... f[t.Undelying] ...  }}} }

```

We keep the `Map[X, Y]` as a whole because we cant recreate a tree for all cases types in the `TypeTree`. In this particular case it seems we could handle it as and `AppliedType` with two captured types but it is not clear if that would be better in general.